.travis.yml

```diff
@@ -1,53 +1,46 @@
-dist: trusty
 sudo: required
 
-language: python
-python:
-  - "3.5.5"
-
 services:
-  - rabbitmq
-  - mysql
+  - docker
 
 env:
-  global:
-    - DBNAME=ottertune
-    - DBUSER=root
-    - WEBSITE_ROOT=server/website
-    - COLON=":"
-    - GRADLE_VERSION=gradle-5.5.1
-    - GRADLE_HOME=/opt/${GRADLE_VERSION}
-    - PATH=${GRADLE_HOME}/bin:${PATH}
+  - DOCKER_COMPOSE_VERSION=1.24.1
 
 before_install:
-  - sudo apt-get -qq update
-  - sudo apt-get install -y python-pip python-dev python-mysqldb python-tk
-  - mysql -e "CREATE DATABASE IF NOT EXISTS ${DBNAME};"
-  - mysql -e "CREATE DATABASE IF NOT EXISTS test_${DBNAME};"
+  # Install latest versions of docker and docker-compose
+  - curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+  - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+  - sudo apt-get update
+  - sudo apt-get -y -o Dpkg::Options::="--force-confnew" install docker-ce
+  - sudo rm /usr/local/bin/docker-compose
+  - curl -L https://github.com/docker/compose/releases/download/${DOCKER_COMPOSE_VERSION}/docker-compose-`uname -s`-`uname -m` > docker-compose
+  - chmod +x docker-compose
+  - sudo mv docker-compose /usr/local/bin
+  - docker --version
+  - docker-compose --version
 
 install:
-  - pip install -r ${WEBSITE_ROOT}/requirements.txt
-  - pip install coverage codecov
-  - wget https://services.gradle.org/distributions/${GRADLE_VERSION}-bin.zip
-  - sudo unzip ${GRADLE_VERSION}-bin.zip -d /opt
-  - gradle --version
-
-before_script:
-  - cp ${WEBSITE_ROOT}/website/settings/credentials_TEMPLATE.py
-       ${WEBSITE_ROOT}/website/settings/credentials.py
-  - sed -i
-        -e "s/^DEBUG.*/DEBUG = True/"
-        -e "s/'USER'${COLON} 'ADD ME\!\!'/'USER'${COLON} '${DBUSER}'/"
-        -e "s/'PASSWORD'${COLON} 'ADD ME\!\!'/'PASSWORD'${COLON} ''/"
-        ${WEBSITE_ROOT}/website/settings/credentials.py
-  - cat ${WEBSITE_ROOT}/website/settings/credentials.py
-  - python ${WEBSITE_ROOT}/manage.py migrate
+  - sudo service mysql stop || true
 
 script:
-  - cd ./client/controller && gradle build
-  - cd ../../server && python -m unittest discover -s analysis/tests -v
-  - cd ../${WEBSITE_ROOT} && coverage run manage.py test --noinput -v 2
-  - cd ../.. && python script/validators/source_validator.py
+  - cd $TRAVIS_BUILD_DIR/docker
+  # Build master images and run the webserver
+  - docker-compose build
+  - docker-compose up -d
+  - docker logs ottertune
+  - docker-compose rm -f -s -v
+  # Build test images and run tests
+  - ci_env=`bash <(curl -s https://codecov.io/env)`
+  - docker-compose -f docker-compose.test.yml build
+  - docker-compose -f docker-compose.test.yml up -d
+  - docker-compose -f docker-compose.test.yml run --workdir="/app/client/controller" --rm test gradle build
```

Comment:
Just an idea: couldn't you add the commands in a bash script (say, test.sh) and then use it as entrypoint of Dockerfile.test (as opposed to Dockefile that has start.sh as entrypoint)? Looking at the output of Travis, it looks like it is starting MySQL for each of the commands from line 36-39. Using the script I cited the lines would go up to line 35.

Reply:
OTOH, I don't know for sure if this test script would introduce bugs in the running of Travis...

Reply:
The issue with putting all of the tests into a single script is that travis will then treat it as a single test instead of 4 individual tests. It's a bit messy but not too bad for 4 tests. If we add more types of tests in the future then we might need to revisit this.

Reply:
Makes total sense. As you can see, my knowledge of Travis falls short, hahah